hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java

```diff
@@ -1412,6 +1428,81 @@ private void appendIfNotEmpty(StringBuilder sb, String regEx,
     }
   }
 
+  /**
+   * Add MD5 hash as checksum request header to the append request
+   * @param requestHeaders
+   * @param reqParams
+   * @param buffer
+   */
+  private void addCheckSumHeaderForWrite(List<AbfsHttpHeader> requestHeaders,
+      final AppendRequestParameters reqParams, final byte[] buffer) {
+    try {
+      MessageDigest md5Digest = MessageDigest.getInstance("MD5");
+      byte[] md5Bytes = md5Digest.digest(
+          Arrays.copyOfRange(buffer, reqParams.getoffset(), reqParams.getLength()));
+      String md5Hash = Base64.getEncoder().encodeToString(md5Bytes);
+      requestHeaders.add(new AbfsHttpHeader(CONTENT_MD5, md5Hash));
+    } catch (NoSuchAlgorithmException e) {
+      e.printStackTrace();
+    }
+  }
+
+  /**
+   * T verify the checksum information received from server for the data read
+   * @param buffer stores the data received from server
+   * @param result HTTP Operation Result
+   * @param bufferOffset Position where data returned by server is saved in buffer
+   * @throws AbfsRestOperationException
+   */
+  private void verifyCheckSumForRead(final byte[] buffer, final AbfsHttpOperation result, final int bufferOffset)
+      throws AbfsRestOperationException{
+    // Number of bytes returned by server could be less than or equal to what
+    // caller requests. In case it is less, extra bytes will be initialized to 0
+    // Server returned MD5 Hash will be computed on what server returned.
+    // We need to get exact data that server returned and compute its md5 hash
+    // Computed hash should be equal to what server returned
+    int numberOfBytesRead = (int)result.getBytesReceived();
+    if (numberOfBytesRead == 0) {
+      return;
+    }
+    byte[] dataRead = new byte[numberOfBytesRead];
+    System.arraycopy(buffer, bufferOffset, dataRead, 0, numberOfBytesRead);
+
+    try {
+      MessageDigest md5Digest = MessageDigest.getInstance(MD5);
+      byte[] md5Bytes = md5Digest.digest(dataRead);
+      String md5HashComputed = Base64.getEncoder().encodeToString(md5Bytes);
+      String md5HashActual = result.getResponseHeader(CONTENT_MD5);
+      if (!md5HashComputed.equals(md5HashActual)) {
+        throw new InvalidChecksumException(null);
+      }
+    } catch (NoSuchAlgorithmException e) {
+      throw new InvalidChecksumException(e);
```

Comment:
Having said above, here no need to handle much. Reason being, it is more of a non-possible exception, since our algo used is constant. If it breaks, it breaks for all. Cant break from some cases.